src/datascience-ui/ipywidgets/kernel/index.ts

```diff
@@ -0,0 +1,263 @@
+class WidgetManagerComponent {
+    private readonly widgetManager: WidgetManager;
+    private readonly scriptManager: ScriptManager;
+    private widgetsCanLoadFromCDN: boolean = false;
+    constructor(private postOffice: PostOffice) {
+        this.scriptManager = new ScriptManager(postOffice);
+        this.scriptManager.onWidgetLoadError(this.handleLoadError.bind(this));
+        this.scriptManager.onWidgetLoadSuccess(this.handleLoadSuccess.bind(this));
+        this.scriptManager.onWidgetVersionNotSupported(this.handleUnsupportedWidgetVersion.bind(this));
+        // tslint:disable-next-line: no-any
+        this.widgetManager = new WidgetManager(undefined as any, postOffice, this.scriptManager.getScriptLoader());
+
+        postOffice.addHandler({
+            // tslint:disable-next-line: no-any
+            handleMessage: (type: string, payload?: any) => {
+                if (type === SharedMessages.UpdateSettings) {
+                    const settings = JSON.parse(payload);
+                    this.widgetsCanLoadFromCDN = settings.widgetScriptSources.length > 0;
+                }
+                return true;
+            }
+        });
+    }
+    public dispose() {
+        this.widgetManager.dispose();
+    }
+    private async handleLoadError(data: {
+        className: string;
+        moduleName: string;
+        moduleVersion: string;
+        // tslint:disable-next-line: no-any
+        error: any;
+        timedout?: boolean;
+        isOnline: boolean;
+    }) {
+        this.postOffice.sendMessage<IInteractiveWindowMapping>(InteractiveWindowMessages.IPyWidgetLoadFailure, {
+            className: data.className,
+            moduleName: data.moduleName,
+            moduleVersion: data.moduleVersion,
+            cdnsUsed: this.widgetsCanLoadFromCDN,
+            isOnline: data.isOnline,
+            timedout: data.timedout,
+            error: data.error
+        });
+        renderErrorInLastOutputThatHasNotRendered(data.error);
+    }
+
+    private handleUnsupportedWidgetVersion(data: { moduleName: 'qgrid'; moduleVersion: string }) {
+        this.postOffice.sendMessage<IInteractiveWindowMapping>(
+            InteractiveWindowMessages.IPyWidgetWidgetVersionNotSupported,
+            {
+                moduleName: data.moduleName,
+                moduleVersion: data.moduleVersion
+            }
+        );
+    }
+
+    private handleLoadSuccess(data: { className: string; moduleName: string; moduleVersion: string }) {
+        this.postOffice.sendMessage<IInteractiveWindowMapping>(InteractiveWindowMessages.IPyWidgetLoadSuccess, {
+            className: data.className,
+            moduleName: data.moduleName,
+            moduleVersion: data.moduleVersion
+        });
+    }
+}
+
+// tslint:disable-next-line no-empty
+const noop = () => {};
+
+const outputDisposables = new Map<string, { dispose(): void }>();
+const outputDisposables2 = new WeakMap<HTMLElement, { dispose(): void }>();
```

Comment:
Sugg:
outputIdDisposables and HTMLElementDisposables or something like that. X and X2 is usually never a useful variable name.